### ISSUE TITLE
Include actual boundaries for paragraphs in rules_block paragraph js

### ISSUE THREAD

**Issue by SamSaffron** [markdown-it/markdown-it#375]
I have a conundrum. 

At Discourse we automatically "embed" links in posts (onebox) 

To allow people to trivially "bypass" the onboxing logic they can precede the link with a space so:

This gets a onebox:

```
http://some.site.com
```

```
abc
http://some.site.com
```

This does not get a onebox: 

```
 http://some.site.com
``` 

Trouble is in my core rule: https://github.com/discourse/discourse/blob/master/app/assets/javascripts/pretty-text/engines/markdown-it/onebox.js.es6 I have no way of telling if there is a space preceding my link_open or not cause `paragraph.js` trims whitespace. 

I am trying to think of ways around this problem and I can think two approaches:

1. Full position mapping in every token, so each token would store `token.pos_map = [start, finish]`, this feels somewhat expensive and complicated to change

2. Amend paragraph.js so it stops trimming, and have text_collapse take care of collapsing the blank tokens. 

I am not sure what the correct way is to go about this, any thoughts?

**Comment by puzrin** [comment]
Probably, i do not understand the problem. Could you rephrase? It's not a big deal to check tokens before/after link (link tokens should be first/last or have breaks before/after).

The only note about your code - add check that link was [autogenerated](https://github.com/markdown-it/markdown-it/blob/8.3.1/lib/rules_core/linkify.js#L105). Because when user type link content explicit, you may wish to keep it intack.

**Comment by SamSaffron** [comment]
problem is the inline token (and paragraph token and link token) do not have any information about where they start in `state.src`

`{"type"=>"inline", "tag"=>"", "attrs"=>nil, "map"=>[0, 1], "nesting"=>0, "level"=>1,`

I can tell that the inline is mapping lines 0 and 1, but 

- I can not tell what actual line my current token is on.
- I can not tell the bMarks of the line in a trivial way short of rescanning. 

I am thinking of 2 changes that would allow me to solve my problem:

1. In paragraph.js instead of running `trim()` on the `state.getLines` store 2 additional tokens (that are not rendered) for leading spaces and trailing spaces in the paragraph. 

2. in newline.js inline rule, instead of "skipping" spaces, store the same psuedo token for "leading" spaces. 

Then I can check for that token in the core rule and if I see it I know I have a leading space before the link.

**Comment by SamSaffron** [comment]
or maybe just have linkify store the `pos` in a meta info on the token?

**Comment by SamSaffron** [comment]
btw... adding a check for `auto` now great catch

**Comment by puzrin** [comment]
AFAIK, linkifier is invoked in the end, after spaces are lost. Seems, the only correct solution is to replace paragraph rule with hacked one, which stores trimming info in token (should be patched [before this line](https://github.com/markdown-it/markdown-it/blob/8.3.1/lib/rules_block/paragraph.js#L35))

**Comment by SamSaffron** [comment]
closing this per:

https://github.com/discourse/discourse/commit/0d62420cbe443b2cf1061405f83f34eef5103532

I am not sure if we need this in markdown.it quite yet, the patch seems to do the trick and tests should catch a regression if we diverge from core. 

Thanks heaps
